src/wallet/wallet.cpp

```diff
@@ -1156,6 +1156,11 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
     if (!strCmd.empty())
     {
         boost::replace_all(strCmd, "%s", wtxIn.GetHash().GetHex());
+#ifdef WIN32
+        boost::replace_all(strCmd, "%w", GetName());
+#else
+        boost::replace_all(strCmd, "%w", ShellEscape(GetName()));
+#endif
```

Comment:
Prefer making `ShellEscape` a noop on WIN32 over conditionals everywhere...

Reply:
You mean identity.

Reply:
Please don't do this is. Please leave behavior **unchanged** on windows in this PR, and do not perform substitution on windows as laanwj suggested https://github.com/bitcoin/bitcoin/pull/13339#discussion_r337990307 and https://github.com/bitcoin/bitcoin/pull/13339#issuecomment-537384875

We are talking about an obscure new feature. Better to not support it on windows, than to implement it an a half baked way where it will sometimes work, sometimes mangle the string it is trying to pass, and sometimes turn on your webcam, upload your private keys, and reset your gmail password.

Reply:
> sometimes mangle the string it is trying to pass, and sometimes turn on your webcam, upload your private keys, and reset your gmail password

if this is the case we should probably have tests that cover these broken scenarios?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/13339#discussion_r357729306

> > sometimes mangle the string it is trying to pass, and sometimes turn on your webcam, upload your private keys, and reset your gmail password
> 
> if this is the case we should probably have tests that cover these broken scenarios?

I'm confused figuring out if this is a joke or a serious suggestion. But to take this straightforwardly, writing a python test to verify that lack of escaping on windows allows running arbitrary shell commands over RPC would just require writing a python test that:

- Starts bitcoind with `-walletnotify="echo %w"`
- Calls `createwallet` with wallet name `& echo It is fun to run shell commands from RPC > funfile`
- Triggering a wallet notification
- Checking that a new file `funfile` exists with contents `It is fun to run shell commands from RPC`

This would (I hope obviously) not be my preferred approach. My preferred approach would be the one suggested by laanwj https://github.com/bitcoin/bitcoin/pull/13339#discussion_r337990307 and https://github.com/bitcoin/bitcoin/pull/13339#issuecomment-537384875 implemented like:

```diff
diff --git a/src/wallet/init.cpp b/src/wallet/init.cpp
index 57b9fdd4c1f..bb18efc9412 100644
--- a/src/wallet/init.cpp
+++ b/src/wallet/init.cpp
@@ -59,7 +59,7 @@ void WalletInit::AddWalletOptions() const
     gArgs.AddArg("-walletbroadcast",  strprintf("Make the wallet broadcast transactions (default: %u)", DEFAULT_WALLETBROADCAST), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     gArgs.AddArg("-walletdir=<dir>", "Specify directory to hold wallets (default: <datadir>/wallets if it exists, otherwise <datadir>)", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
 #if HAVE_SYSTEM
-    gArgs.AddArg("-walletnotify=<cmd>", "Execute command when a wallet transaction changes (%s in cmd is replaced by TxID and %w is replaced by wallet name)", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
+    gArgs.AddArg("-walletnotify=<cmd>", "Execute command when a wallet transaction changes. %s in cmd is replaced by TxID and %w is replaced by wallet name. %w is not currently implmented on windows. On systems where %w is supported, it should NOT be quoted because this would break shell escaping used to invoke the command.", ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
 #endif
     gArgs.AddArg("-walletrbf", strprintf("Send transactions with full-RBF opt-in enabled (RPC only, default: %u)", DEFAULT_WALLET_RBF), ArgsManager::ALLOW_ANY, OptionsCategory::WALLET);
     gArgs.AddArg("-zapwallettxes=<mode>", "Delete all wallet transactions and only recover those parts of the blockchain through -rescan on startup"
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index a8ae5ef62c4..ad0a918d7aa 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1156,9 +1156,12 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
     if (!strCmd.empty())
     {
         boost::replace_all(strCmd, "%s", wtxIn.GetHash().GetHex());
-#ifdef WIN32
-        boost::replace_all(strCmd, "%w", GetName());
-#else
+#ifndef WIN32
+        // Substituting the wallet name isn't currently supported on windows
+        // because windows shell escaping has not been implemented yet:
+        // https://github.com/bitcoin/bitcoin/pull/13339#issuecomment-537384875
+        // A few ways it could be implemented in the future are described in:
+        // https://github.com/bitcoin/bitcoin/pull/13339#issuecomment-461288094
         boost::replace_all(strCmd, "%w", ShellEscape(GetName()));
 #endif
         std::thread t(runCommand, strCmd);
```

Reply:
you're describing a different issue than what I was thinking of, could you have a wallet name that would get passed incorrectly if it was not escaped properly? for instance with a wallet named `wallet`, `wallet 1`, `wallet 2`. I'm just trying to understand all the different ways the current code would break, and suggested tests that document this, so no I was not joking.

Reply:
Regardless, I believe that you're right that this should be a noop if we're not supporting it on windows...

Reply:
Ah I see. I hope that comment didn't come off wrong. I believe an example of a name that'd be passed incorrectly would be a wallet named `eggs & bacon`. With cacc63bed101ecb56679695e1b490fd76f83bd50 and `walletnotify=walletnotify.exe %w`, I think the notifications script would see command line string `walletnotify.exe 'eggs`, because the `cmd.exe` command interpreter on windows treats `&` as a separator between commands, and doesn't treat single quotes specially (only double quotes).